src/dwarf.rs

```diff
@@ -966,6 +1122,305 @@ impl DwarfResolver {
+                if let Some(sibling_off) = find_die_sibling(&mut die) {
+                    dieiter.seek_to_sibling(sibling_off);
+                    continue;
+                }
+                // Skip this DIE quickly, or the iterator will
+                // recalculate the size of the DIE.
+                die.exhaust().unwrap();
+            }
+            continue;
+        }
+
+        if let Ok(syminfo) = parse_die_subprogram(&mut die, str_data) {
+            if syminfo.is_some() {
+                found_syms.push(syminfo.unwrap());
+            }
+        }
+    }
+}
+
+/// The parse result of the `.debug_info` section.
+///
+/// This type is used by the worker threads to pass results to the
+/// coordinator after finishing an Unit.  `Stop` is used to nofity the
+/// coordinator that a matching condition is met.  It could be that
+/// the given symbol is already found, so that the coordinator should
+/// stop producing more tasks.
+enum DIParseResult<'a> {
+    Symbols(Vec<DWSymInfo<'a>>),
+    Stop,
+}
+
+/// Parse the addresses of symbols from the `.debug_info` section.
+///
+/// # Arguments
+///
+/// * `parser` - is an ELF parser.
+/// * `cond` - is a function to check if we have found the information
+///            we need.  The function will stop earlier if the
+///            condition is met.
+/// * `nthreads` - is the number of worker threads to create. 0 or 1
+///                means single thread.
+fn debug_info_parse_symbols<'a>(
+    parser: &'a Elf64Parser,
+    cond: Option<&(dyn Fn(&DWSymInfo<'a>) -> bool + Send + Sync)>,
+    nthreads: usize,
+) -> Result<Vec<DWSymInfo<'a>>, Error> {
+    let info_sect_idx = parser.find_section(".debug_info")?;
+    let info_data = parser.read_section_raw_cache(info_sect_idx)?;
+    let abbrev_sect_idx = parser.find_section(".debug_abbrev")?;
+    let abbrev_data = parser.read_section_raw_cache(abbrev_sect_idx)?;
+    let units = debug_info::UnitIter::new(&info_data, &abbrev_data);
+    let str_sect_idx = parser.find_section(".debug_str")?;
+    let str_data = parser.read_section_raw_cache(str_sect_idx)?;
+
+    #[cfg(feature = "dwarf_perf")]
+    let now = Instant::now();
+
+    let mut syms = Vec::<DWSymInfo>::new();
+
+    if nthreads > 1 {
+        // Create worker threads to process tasks (Units) in a work
+        // queue.
+        let mut handles = vec![];
+        let mut queue = SyncQueue::<debug_info::DIEIter<'static>>::new();
+        let cond = unsafe {
+            mem::transmute::<_, Option<&'static (dyn Fn(&DWSymInfo) -> bool + Send + Sync)>>(cond)
+        };
+        let (result_tx, result_rx) = mpsc::channel::<DIParseResult>();
+
+        for _ in 0..nthreads {
+            let result_tx = result_tx.clone();
+            let str_data = unsafe { mem::transmute::<_, &'static [u8]>(str_data) };
+            let queue = unsafe {
+                mem::transmute::<_, &mut SyncQueue<debug_info::DIEIter<'static>>>(&mut queue)
+            };
```

Comment:
see my comment below about https://crates.io/crates/crossbeam-channel, all this transmutes and unsafe to use the queue seems like a big red flag

Reply:
I just checked the API thread::scope().  It will solve the problem of unsafe code here.